ibis/impala/compiler.py

```diff
@@ -181,6 +185,111 @@ def _cumulative_to_window(translator, expr, window):
     return new_expr
 
 
+_window_types = {
+    'row': 'ROWS',
+    'range': 'RANGE'
+}
+
+
+_map_interval_to_microseconds = dict(
+    W=6.048e11,
+    D=8.64e10,
+    h=3.6e9,
+    m=6e7,
+    s=1e6,
+    ms=1000,
+    us=1,
+    ns=.001
+)
+
+
+_map_interval_op_to_op = {
+    # Literal Intervals have two args, i.e.
+    # Literal(1, Interval(value_type=int8, unit='D', nullable=True))
+    # Parse both args and multipy 1 * _map_interval_to_microseconds['D']
+    ops.Literal: lambda x, y: x * y,
+    ops.IntervalMultiply: lambda x, y: x * y,
+    ops.IntervalAdd: lambda x, y: x + y,
+    ops.IntervalSubtract: lambda x, y: x - y,
+}
+
+
+def _replace_interval_with_scalar(expr):
+    """
+    Good old Depth-First Search to identify the Interval and IntervalValue
+    components of the expression and return a comparable scalar expression.
+
+    Parameters
+    ----------
+    expr : float or expression of intervals, i.e.
+      1 * ibis.day() + 5 * ibis.hour()
+
+    Returns
+    -------
+    preceding : float or ir.FloatingScalar, depending upon the expr
+    """
+    if not isinstance(expr, (dt.Interval, ir.IntervalValue)):
+        try:
+            # Literal expressions have op method but native types do not.
+            if isinstance(expr.op(), ops.Literal):
+                return expr.op().value
+        except AttributeError:
+            return expr
+    elif isinstance(expr, dt.Interval):
+        try:
+            microseconds = _map_interval_to_microseconds[expr.unit]
+            return microseconds
+        except KeyError:
+            raise KeyError(
+                "Expected preceding values of week(), " +
+                "day(), hour(), minute(), second(), millisecond(), " +
+                "microseconds(), nanoseconds(); got {}".format(expr)
+                )
+    elif expr.op().args and isinstance(expr, ir.IntervalValue):
+        if len(expr.op().args) > 2:
+            raise com.NotImplementedError(
+                "'preceding' argument cannot be parsed."
+            )
+        left_arg = _replace_interval_with_scalar(expr.op().args[0])
+        right_arg = _replace_interval_with_scalar(expr.op().args[1])
+        method = _map_interval_op_to_op[type(expr.op())]
```

Comment:
You're using `expr.op()` in a bunch of places, can you make that into a variable?